File: opensource/kernel/src/main/java/com/sos/scheduler/engine/kernel/MavenProperties.java
package com.sos.scheduler.engine.kernel;

import com.sos.scheduler.engine.kernel.util.ClassResource;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;


public final class MavenProperties {
    private static final String resourceSimpleName = "maven.properties";
    private final Properties properties;


    public MavenProperties(ClassResource r) {
        try {
            InputStream in = r.getInputStream();
            try {
                properties = new Properties();
                properties.load(in);
            }
            finally { in.close(); }
        }
        catch (IOException x) { throw new RuntimeException(x); }
    }


    public MavenProperties(Class<?> clas) {
        this(new ClassResource(clas, resourceSimpleName));
    }


    public String getGroupId() {
        return get("project.groupId");
    }
    

    public String getArtifactId() {
        return get("project.artifactId");
    }


    public String getVersion() {
        return get("project.version");
    }


    public String get(String name) {
        String result = properties.getProperty(name);
        if (result == null) throw new RuntimeException("Unknown property '" + name + "'");
        return result;
    }


    @Override public String toString() {
        return getGroupId() + "/" + getArtifactId() + " (" + getVersion() + ")";
    }
}

File: cppjava/generator/src/main/scala/com/sos/scheduler/engine/cplusplus/generator/main/MavenProperties.scala
package com.sos.scheduler.engine.cplusplus.generator.main

import com.sos.scheduler.engine.cplusplus.scalautil.io.Util._

class MavenProperties(clas: Class[_]) {
  private val resourceSimpleName = "maven.properties"
  def groupId: String = get("project.groupId")
  def artifactId: String = get("project.artifactId")
  def version: String = get("project.version")

  def get(name: String) = getOption(name) getOrElse { throw new RuntimeException("Unknown property '" + name + "'") }

  def getOption(name: String) = Option(properties.getProperty(name))

  lazy val properties = {
    val resourcePath = clas.getName.split('.').dropRight(1).mkString("/", "/", "/") + resourceSimpleName
    val result = new java.util.Properties
    closingFinally(openResource(resourcePath)) { in => result.load(in) }
    result
  }

  private def openResource(path: String) = Option(clas.getResourceAsStream(path)) getOrElse sys.error("Missing Java resource '" + path + "'")

  override def toString = groupId + "/" + artifactId + " (" + version + ")"
}
